vpn/tls_test.go

```diff
@@ -480,10 +487,157 @@ func Test_tlsHandshake(t *testing.T) {
 		t.Errorf("tlsHandshake() error = %v, wantErr %v", err, wantErr)
 		return
 	}
+
+	// we bubble up any error coming from the factory
+	tlsFactoryFn = errorRaisingTLSFactory
+	wantErr = tlsFactoryError
+	_, err = tlsHandshake(conn, conf)
+	if !errors.Is(err, wantErr) {
+		t.Errorf("tlsHandshake() error = %v, wantErr %v", err, wantErr)
+		return
+	}
 }
 
 func Test_defaultTLSFactory(t *testing.T) {
 	conn := &mocks.Conn{}
 	conf := &tls.Config{}
 	defaultTLSFactory(conn, conf)
 }
+
+func Test_parrotTLSFactory(t *testing.T) {
+	conn := &mocks.Conn{}
+	conf := &tls.Config{InsecureSkipVerify: true}
+	_, err := parrotTLSFactory(conn, conf)
+	if err != nil {
+		t.Errorf("parrotTLSFactory() error = %v, wantErr %v", err, nil)
+		return
+	}
+
+	// an hex clienthello that cannot be decoded to raw bytes should raise ErrBadParrot
+	vpnClientHelloHex = `aaa`
+	_, err = parrotTLSFactory(conn, conf)
+	wantErr := ErrBadParrot
+	if !errors.Is(err, wantErr) {
+		t.Errorf("tlsHandshake() error = %v, wantErr %v", err, wantErr)
+		return
+	}
+
+	// an hex representation that is not a valid clienthello should raise ErrBadParrot
+	vpnClientHelloHex = `deadbeef`
+	_, err = parrotTLSFactory(conn, conf)
+	wantErr = ErrBadParrot
+	if !errors.Is(err, wantErr) {
+		t.Errorf("tlsHandshake() error = %v, wantErr %v", err, wantErr)
+		return
+	}
+
+	// TODO(ainghazal): there's an extra error case that I'm not pretty sure how to reach
+	// (error on client.ApplyPreset)
+}
+
+func Test_customVerify(t *testing.T) {
+
+	// a correct certChain should validate
+
+	rawCerts, err := makeRawCerts()
+	if err != nil {
+		t.Errorf("error getting raw certs")
+		return
+	}
+
+	err = customVerify(rawCerts, nil)
+	if err != nil {
+		t.Errorf("customVerify() error = %v, wantErr %v", err, nil)
+	}
+
+	// empty certchain raises error
+
+	emptyCerts := [][]byte{[]byte{}, []byte{}}
+	wantErr := ErrCannotVerifyCertChain
+	err = customVerify(emptyCerts, nil)
+	if !errors.Is(err, wantErr) {
+		t.Errorf("customVerify() error = %v, wantErr %v", err, wantErr)
+	}
+
+	// garbage certchain raises error
+
+	garbageCerts := [][]byte{[]byte{0xde, 0xad}, []byte{0xbe, 0xef}}
+	wantErr = ErrCannotVerifyCertChain
+	err = customVerify(garbageCerts, nil)
+	if !errors.Is(err, wantErr) {
+		t.Errorf("customVerify() error = %v, wantErr %v", err, wantErr)
+	}
+
+	// attempting to verify one cert with a different ca raises error
+
+	certChainOne, _ := makeRawCerts()
+	certChainTwo, _ := makeRawCerts()
+	badChain := [][]byte{certChainOne[0], certChainTwo[1]}
+	wantErr = ErrCannotVerifyCertChain
+	err = customVerify(badChain, nil)
+	if !errors.Is(err, wantErr) {
+		t.Errorf("customVerify() error = %v, wantErr %v", err, wantErr)
+	}
+}
+
+// makeRawCerts creates a CA, and return an array of byte arrays containing a
+// cert signed with that CA and the CA itself. it also returns an error if it
+// could not build the certs correctly.
+func makeRawCerts() ([][]byte, error) {
```

Comment:
You could consider using github.com/google/martian to generate a CA. It contains a subpackage, [mitm](https://pkg.go.dev/github.com/google/martian/v3/mitm), that should simplify generating a CA.